pkg/logging/grpc.go

```diff
@@ -0,0 +1,208 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package logging
+
+import (
+	"fmt"
+	"math/rand"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors"
+	grpc_logging "github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/logging"
+	"github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/tags"
+	"github.com/oklog/ulid"
+	"google.golang.org/grpc/status"
+	"gopkg.in/yaml.v2"
+)
+
+// NewRequestConfig parses the string into a req logging config structure.
+// Raise an error if unmarshalling is not possible, or values are not valid.
+func NewRequestConfig(configYAML []byte) (*RequestConfig, error) {
+	reqLogConfig := &RequestConfig{}
+	if err := yaml.UnmarshalStrict(configYAML, reqLogConfig); err != nil {
+		return nil, err
+	}
+	return reqLogConfig, nil
+}
+
+// Checks if the OptionsConfig struct is empty and valid.
+// If invalid combination is present, return an error.
+func checkOptionsConfigEmpty(optcfg OptionsConfig) (bool, error) {
+	if len(optcfg.Level) == 0 && !optcfg.Decision.LogEnd && !optcfg.Decision.LogStart {
+		return true, nil
+	}
+	if len(optcfg.Level) == 0 && (optcfg.Decision.LogStart || optcfg.Decision.LogEnd) {
+		return false, fmt.Errorf("level field is empty.")
+	}
+	return false, nil
+}
+
+// This function configures all the method to have global config for logging.
+func fillGlobalOptionConfig(reqLogConfig *RequestConfig, isgRPC bool) (string, bool, bool, error) {
+	globalLevel := "ERROR"
+	globalStart, globalEnd := false, false
+
+	globalOptionConfig := reqLogConfig.Options
+	isEmpty, err := checkOptionsConfigEmpty(globalOptionConfig)
+
```

Comment:
Let's not introduce empty lines before err checking - it's crucial to expect err check just below (TODO: Put in style guide)

Reply:
Sure